File: src/main/java/org/lanternpowered/server/block/tile/vanilla/ContainerBlockEntityBase.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.block.tile.vanilla;

import org.lanternpowered.server.block.tile.LanternBlockEntity;
import org.lanternpowered.server.block.vanilla.container.action.ContainerAnimationAction;
import org.lanternpowered.server.inventory.AbstractContainer;
import org.lanternpowered.server.inventory.InventoryViewerListener;
import org.lanternpowered.server.world.LanternWorld;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.world.Location;

import java.util.HashSet;
import java.util.Set;

public abstract class ContainerBlockEntityBase extends LanternBlockEntity implements InventoryViewerListener {

    protected final Set<Player> viewers = new HashSet<>();

    /**
     * The delay that will be used to play
     * the open/close sounds.
     */
    private int soundDelay;

    @Override
    public void onViewerAdded(Player viewer, AbstractContainer container, Callback callback) {
        if (this.viewers.add(viewer) && this.viewers.size() == 1) {
            this.soundDelay = getOpenSoundDelay();

            final Location location = getLocation();
            final LanternWorld world = (LanternWorld) location.getWorld();
            world.addBlockAction(location.getBlockPosition(), getBlock().getType(), ContainerAnimationAction.OPEN);
        }
    }

    @Override
    public void onViewerRemoved(Player viewer, AbstractContainer container, Callback callback) {
        if (this.viewers.remove(viewer) && this.viewers.size() == 0) {
            this.soundDelay = getCloseSoundDelay();

            final Location location = getLocation();
            final LanternWorld world = (LanternWorld) location.getWorld();
            world.addBlockAction(location.getBlockPosition(), getBlock().getType(), ContainerAnimationAction.CLOSE);
        }
    }

    /**
     * Gets the delay that should be used to
     * play the open sound.
     *
     * @return The open sound delay
     */
    protected int getOpenSoundDelay() {
        return 5;
    }

    /**
     * Gets the delay that should be used to
     * play the open sound.
     *
     * @return The open sound delay
     */
    protected int getCloseSoundDelay() {
        return 10;
    }

    /**
     * Plays the open sound at the {@link Location}.
     *
     * @param location The location
     */
    protected abstract void playOpenSound(Location location);

    /**
     * Plays the close sound at the {@link Location}.
     *
     * @param location The location
     */
    protected abstract void playCloseSound(Location location);

    @Override
    public void pulse() {
        super.pulse();

        if (this.soundDelay > 0 && --this.soundDelay == 0) {
            final Location location = getLocation();
            if (this.viewers.size() > 0) {
                playOpenSound(location);
            } else {
                playCloseSound(location);
            }
        }
    }
}

File: src/main/java/org/lanternpowered/server/block/entity/vanilla/ContainerBlockEntityBase.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.block.entity.vanilla

import org.lanternpowered.api.world.Location
import org.lanternpowered.server.block.entity.BlockEntityCreationData
import org.lanternpowered.server.block.entity.LanternBlockEntity
import org.lanternpowered.server.block.vanilla.container.action.ContainerAnimationAction
import org.lanternpowered.server.inventory.AbstractContainer
import org.lanternpowered.server.inventory.InventoryViewerListener
import org.lanternpowered.server.world.LanternWorldNew
import org.spongepowered.api.entity.living.player.Player
import java.util.HashSet
import kotlin.time.Duration
import kotlin.time.seconds

abstract class ContainerBlockEntityBase(creationData: BlockEntityCreationData) :
        LanternBlockEntity(creationData), InventoryViewerListener {

    protected val viewers = HashSet<Player>()

    /**
     * The delay that will be used to play the open/close sounds.
     */
    private var soundDelay = 0.seconds

    /**
     * Gets the delay that should be used to
     * play the open sound.
     *
     * @return The open sound delay
     */
    protected val openSoundDelay: Duration
        get() = 0.25.seconds

    /**
     * Gets the delay that should be used to
     * play the open sound.
     *
     * @return The open sound delay
     */
    protected val closeSoundDelay: Duration
        get() = 0.5.seconds

    override fun onViewerAdded(viewer: Player, container: AbstractContainer, callback: InventoryViewerListener.Callback) {
        if (!this.viewers.add(viewer) || this.viewers.size != 1)
            return
        this.soundDelay = this.openSoundDelay
        val location = this.location
        val world = location.world as LanternWorldNew
        world.addBlockAction(location.blockPosition, this.block.type, ContainerAnimationAction.OPEN)
    }

    override fun onViewerRemoved(viewer: Player, container: AbstractContainer, callback: InventoryViewerListener.Callback) {
        if (!this.viewers.remove(viewer) || this.viewers.size != 0)
            return
        this.soundDelay = this.closeSoundDelay
        val location = this.location
        val world = location.world as LanternWorldNew
        world.addBlockAction(location.blockPosition, block.type, ContainerAnimationAction.CLOSE)
    }

    /**
     * Plays the open sound at the [Location].
     *
     * @param location The location
     */
    protected abstract fun playOpenSound(location: Location)

    /**
     * Plays the close sound at the [Location].
     *
     * @param location The location
     */
    protected abstract fun playCloseSound(location: Location)

    override fun update(deltaTime: Duration) {
        super.update(deltaTime)

        if (this.soundDelay > Duration.ZERO) {
            this.soundDelay -= deltaTime
            if (this.soundDelay <= Duration.ZERO) {
                val location = this.location
                if (this.viewers.size > 0) {
                    this.playOpenSound(location)
                } else {
                    this.playCloseSound(location)
                }
            }
        }
    }
}
